src/java.base/share/classes/java/util/concurrent/LinkedTransferQueue.java

```diff
@@ -528,73 +557,59 @@ final int addSweepVote() {
-                    haveData != ((item = p.item) != null) &&
-                    p.cmpExItem(item, e) == item) {
-                    Thread w = p.waiter;
-                    if (slack) {
-                        DualNode n = ((q = p.next) == null) ? p : q;
-                        if (h != n && h == cmpExHead(h, n))
-                            h.selfLink();    // advance by 2 if possible
-                    }
+            p = (t = tail) != null && t.isData == haveData && t != prevp ? t : h;
+            prevp = p;                      // avoid known self-linked tail path
+            do {
+                m = p.item;
+                q = p.next;
+                if (p.isData != haveData && haveData != (m != null) &&
+                    p.cmpExItem(m, e) == m) {
+                    Thread w = p.waiter;    // matched complementary node
+                    if (p != h && h == cmpExHead(h, (q == null) ? p : q))
+                        h.next = h;         // advance head; self-link old
                     LockSupport.unpark(w);
-                    return item;
-                }
-                if ((q = p.next) == null) {  // try to append
-                    if (nanos == 0L)         // unless immediate
-                        return e;
-                    if (s == null) {
+                    return m;
+                } else if (q == null) {
+                    if (ns == 0L)           // try to append unless immediate
+                        break restart;
+                    if (s == null)
                         s = new DualNode(e, haveData);
-                        q = p.next;          // recheck after allocation
-                    }
-                    if (q == null && (q = p.cmpExNext(null, s)) == null) {
-                        if (slack)
+                    if ((q = p.cmpExNext(null, s)) == null) {
+                        if (p != t)
                             cmpExTail(t, s);
-                        if (nanos < 0L)
-                            return e;
-                        pred = p;
                         break restart;
                     }
                 }
-                if (p == (p = q))            // self-linked; restart
-                    break;
-            }
+            } while (p != (p = q));         // restart if self-linked
         }
-        Object match = s.await(e, nanos, this, // spin if at or near head
-                               pred == null || pred.waiter == null);
-        if (match == e)
-            unsplice(pred, s);               // cancelled
-        return match;
+        if (s == null || ns <= 0L)
+            m = e;                          // don't wait
+        else if ((m = s.await(e, ns, this,  // spin if at or near head
+                              p == null || p.waiter == null)) == e)
+            unsplice(p, s);                 // cancelled
+        else if (m != null)
+            s.selfLinkItem();
+        return m;
```

Comment:
@DougLea I'd probably add a newline before the return statement to visually distinguish that it isn't intended to be read as a part of the if-else branches.

Reply:
Thanks. Done.